File: Day279_2025/Q279_Leetcode.py
import heapq
from typing import List

class Solution:
    def swimInWater(self, grid: List[List[int]]) -> int:
        n = len(grid)
        pq = [(grid[0][0], 0, 0)]
        seen = set()
        directions = [(0,1), (1,0), (0,-1), (-1,0)]
        
        while pq:
            time, r, c = heapq.heappop(pq)
            if (r, c) in seen:
                continue
            seen.add((r, c))
            
            if r == n - 1 and c == n - 1:
                return time
            
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < n and 0 <= nc < n and (nr, nc) not in seen:
                    heapq.heappush(pq, (max(time, grid[nr][nc]), nr, nc))
        
        return -1

File: Day279_2025/Q279_Leetcode.java
import java.util.*;

class Q279_Leetcode {
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        boolean[][] seen = new boolean[n][n];
        int[][] dirs = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        
        pq.offer(new int[]{grid[0][0], 0, 0});
        
        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            int time = curr[0], r = curr[1], c = curr[2];
            
            if (seen[r][c]) continue;
            seen[r][c] = true;
            
            if (r == n - 1 && c == n - 1) return time;
            
            for (int[] d : dirs) {
                int nr = r + d[0], nc = c + d[1];
                if (nr >= 0 && nr < n && nc >= 0 && nc < n && !seen[nr][nc]) {
                    pq.offer(new int[]{Math.max(time, grid[nr][nc]), nr, nc});
                }
            }
        }
        return -1;
    }
}
